### PR TITLE
Authenticate self-hosted sites using Application Password

### PR DESCRIPTION
**Pull request by crazytonyli**
There are two changes in this PR.

### 1. Authenticate self-hosted sites using Application Password

Added a new flow to authenticate self-hosted sites using [application password](https://make.wordpress.org/core/2020/11/05/application-passwords-integration-guide/). This new flow is hidden behind a new local feature flag "Authenticate self-hosted sites using Application Password". At the moment, it's only accessible from site picker.

#### Application password vs blog account password

At the moment, the app only supports actual blog account passwords. It's used, along with username, to authenticate XMLRPC calls.

The key difference to the app between these two passwords is: Application password can be used to call REST API and XMLRPC, but blog account password can only be used in XMLRPC.

The app stores the two difference password in different places: Blog account passwords are [associated with XMLRPC endpoint URL](https://github.com/wordpress-mobile/WordPress-iOS/blob/954bb4cc016f4a7b6fb76af0fd9d40e66442cb43/WordPress/Classes/Models/Blog.m#L485-L490), which is an existing behaviour and hasn't been changed. The new application passwords are associated with site URL.

Please note, in the new flow where user authenticates using application password, the application password will also be stored as an "blog account password". This is so that we don't have to modify the existing code to retrieve password from the new place where application password is stored.

### 2. New screen to manage application passwords

This feature is not accessible at the moment.

## Regression Notes
1. Potential unintended areas of impact

2. What I did to test those areas of impact (or what existing automated tests I relied on)

3. What automated tests I added (or what prevented me from doing so)

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

Testing checklist:
- [ ] WordPress.com sites and self-hosted Jetpack sites.
- [ ] Portrait and landscape orientations.
- [ ] Light and dark modes.
- [ ] Fonts: Larger, smaller and bold text.
- [ ] High contrast.
- [ ] VoiceOver.
- [ ] Languages with large words or with letters/accents not frequently used in English.
- [ ] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [ ] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          3 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false"><code>Modules/Package.swift</code> was changed without updating its corresponding <code>Package.resolved</code>. Please resolve the Swift packages in Xcode.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot or video is included in the pull request. Consider adding some for clarity.</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 500 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WordPress/Classes/Networking/WordPressClient.swift
Comment (crazytonyli) [inline WordPress/Classes/Networking/WordPressClient.swift:29]: With this change, `WordPressSite` and `WordPressClient` would only be accessible to WP.com sites and self-hosted sites that's authenticated using the new application password flow. That means in places that store them as properties, they'll be declared as optional, until we entirely move away from XMLRPC.

@jkmassel Let me know if that deviates from how you originally designed them or you have any concerns.

**Comment by kean** [comment]
> New screen to manage application passwords

This part is pretty rough. I'd be great to review/update it separately.

File: WordPress/Classes/Networking/LoginClient.swift
Comment (kean) [inline WordPress/Classes/Networking/LoginClient.swift]: This might be a retain cycle. It's also unclear why `pushViewController` is followed by the `loginViewController.dismiss` call .

File: WordPress/Classes/Networking/LoginClient.swift
Comment (kean) [inline WordPress/Classes/Networking/LoginClient.swift]: FYI I'm bumping the deployment target here https://github.com/wordpress-mobile/WordPress-iOS/pull/23476. But I couldn't get the CI to build it.

File: WordPress/Classes/Networking/LoginClient.swift
Comment (kean) [inline WordPress/Classes/Networking/LoginClient.swift]: Is `#if os(macOS)` needed?

File: WordPress/Classes/Networking/LoginClient.swift
Comment (kean) [inline WordPress/Classes/Networking/LoginClient.swift:146]: I would suggest implementing a simple `LoginViewController` and use `ASWebAuthenticationSession` and `WordPressAuthenticator.shared.delegate` directly from it.  UIKit will probably be a better option too. Let's not pretend it is a ViewModel. It makes it harder to follow this code.

File: WordPress/Classes/Networking/WordPressClient.swift
Comment (kean) [inline WordPress/Classes/Networking/WordPressClient.swift:35]: What's the role of `WordPressClient`?

File: Modules/Sources/DesignSystem/Components/DSEditableListDetailItem.swift
Comment (kean) [inline Modules/Sources/DesignSystem/Components/DSEditableListDetailItem.swift:3]: Regarding the `DS` part, let's not extend the `DesignSystem` framework and let's not use the `DS` prefix moving forward. I was planning to remove the `DesignSystem` framework along with all of its APIs. The `DSEditableListDetailItemView` and  `DSListDetailItem` components are also arguably too simply to be defined as shared components. They look like specific implementations needed for the `ApplicationTokenListView`.

**Review by kean — COMMENTED** [review]
Great stuff! I ran a smoke test and was able to login.

This screen is not particularly user-friendly. I wonder if we have any control over it. I would get rid of the header and the "you will be redirected to" message - weird.

Screenshot 2024-08-05 at 4 03 03 PM

If you remove and re-add the same site, it uses the same "Password Name" and throws an error "The password with this name already exists". I would suggest adding a UUID or a timestamp to a password: "jetpack-app-1722890146".

It will now work if you have some login-related plugins, like plugins for 2FA, right?

**Comment by crazytonyli** [comment]
> This screen is not particularly user-friendly. I wonder if we have any control over it. I would get rid of the header and the "you will be redirected to" message - weird.

I don't think we have much control over the web UI from a client perspective.

> I would suggest adding a UUID or a timestamp to a password: "jetpack-app-1722890146".

Sounds good. I can add a timestamp at the end.

> It will now work if you have some login-related plugins, like plugins for 2FA, right?

Having 2FA should be okay. Application password authentication may be disabled on the site, with or without plugins (see the "Feature Availability" section of [this post](https://make.wordpress.org/core/2020/11/05/application-passwords-integration-guide/)). In that case, we won't be able to add the site to the app. We can potentially improve the error handling part to show more detailed messaged around what application password is.

File: WordPress/Classes/Networking/LoginClient.swift
Comment (crazytonyli) [reply WordPress/Classes/Networking/LoginClient.swift]: Good catch! Fixed in https://github.com/wordpress-mobile/WordPress-iOS/pull/23473/commits/d8e1b1e99fbdfa910637638291c5c60da6a23a7b

File: WordPress/Classes/Networking/LoginClient.swift
Comment (crazytonyli) [reply WordPress/Classes/Networking/LoginClient.swift]: Don't think so. Removed in https://github.com/wordpress-mobile/WordPress-iOS/pull/23473/commits/0fe3ea7512a61905583c99cf254e5afa1d5ed8ae

File: WordPress/Classes/Networking/LoginClient.swift
Comment (crazytonyli) [reply WordPress/Classes/Networking/LoginClient.swift:146]: I added some code here to make the end-to-end flow work. I plan to refactor this part in a separate PR.

File: WordPress/Classes/Networking/WordPressClient.swift
Comment (crazytonyli) [reply WordPress/Classes/Networking/WordPressClient.swift:35]: A wrapper of the REST API client implemented in wordpress-rs. @jkmassel would be a better person to answer this in detail.

File: Modules/Sources/DesignSystem/Components/DSEditableListDetailItem.swift
Comment (crazytonyli) [reply Modules/Sources/DesignSystem/Components/DSEditableListDetailItem.swift:3]: I'm not aware of our plan for `DesignSystem`, and what should/shouldn't go into it. @jkmassel Do you mind having a look? Should the new SwiftUI views be kept in the `DesignSystem` module, or somewhere else?

**Comment by kean** [comment]
> I don't think we have much control over the web UI from a client perspective.

This extra information is probably there because I'm running the development environment. I think we should be good in prod.

Approving as a pre-production version. I would wait for Jeremy's review as well before merging.

Note that I haven't reviewed the "New screen to manage application passwords". I would appreciate if it could be extracted to a separate PR.

**Review by kean — APPROVED** [review]

**Comment by crazytonyli** [comment]
> Note that I haven't reviewed the "New screen to manage application passwords". I would appreciate if it could be extracted to a separate PR.

The new screen doesn't work anyways. 😄 The loading data needs some changes, which will be in a separate PR.

File: WordPress/Classes/Networking/WordPressClient.swift
Comment (jkmassel) [reply WordPress/Classes/Networking/WordPressClient.swift:29]: > With this change, WordPressSite and WordPressClient would only be accessible to WP.com sites and self-hosted sites that's authenticated using the new application password flow.

This is accurate 

> That means in places that store them as properties, they'll be declared as optional, until we entirely move away from XMLRPC.

I'm not too worried about the _how_ but I'd really like to limit optionality. That's why I implemented `static func from(blog: Blog) throws -> WordPressSite` – I figure we don't need to store this type just yet, we can derive it from wherever we pass a `Blog` (or some way to get a `Blog`) around. 

Long-term we'll have some SwiftUI hierarchy where we inject the blog at the appropriate level and all subcomponents can just reference it but until then I think this derivation approach should scale pretty well?

File: WordPress/Classes/Networking/WordPressClient.swift
Comment (jkmassel) [reply WordPress/Classes/Networking/WordPressClient.swift:35]: IMHO the idea would be that it could encapsulate client access details for a specific site. It would answer questions like:

- Can we currently reach this site via the network? (if not, we should display an offline banner)
- Which radio are we using to access this site? (controls the extent to which we do preloading)
- Does the user have permission to perform `$OPERATION` on this site? (it'd cache this, but could reload the permissions on a force-refresh)

Possibly other stuff – I don't want it to become the everything object, but I also don't want to thread 10 different objects through the object hierarchy either.

**Review by jkmassel — APPROVED** [review]
This seems fine, but I agree with should tear out the `ViewLayer` stuff – mostly because we ended up deciding not to make that a module and go all-in on WordPressUI instead. Would be good to be consistent with that decision.

**Pull request merged**
